cloudinit/sources/__init__.py

```diff
@@ -480,18 +488,19 @@ def get_data(self) -> bool:
 
         Minimally, the datasource should return a boolean True on success.
         """
-        self._dirty_cache = True
-        return_value = self._check_and_get_data()
-        # TODO: verify that datasource types are what they are expected to be
-        # each datasource uses different logic to get userdata, metadata, etc
-        # and then the rest of the codebase assumes the types of this data
-        # it would be prudent to have a type check here that warns, when the
-        # datatype is incorrect, rather than assuming types and throwing
-        # exceptions later if/when they get used incorrectly.
-        if not return_value:
+        with log.log_time("Getting metadata"):
```

Comment:
This is a good addition, but I do wonder about in child classes where we were measuring something narrow in `_get_data()` that now may be much broader. In particular, I think it's useful knowing the time it took to crawl an IMDS without the surrounding processing of _get_data(). I think it's fine as-is, but it might be something we need to further tweak in the future.

Reply:
> I think it's useful knowing the time it took to crawl an IMDS without the surrounding processing of _get_data()

Agreed. I went back and forth on this, and a previous iteration actually added a new standard `crawl_metadata()` to the base class and adapted other datasources to track this seperate from `get_data()`'s time. Ultimately though, I went for a simpler approach because I think that in most cases any extra time is already captured in the log: any datasources that need to set up ephemeral networks already get that captured via `subp`'s timer, and the data manipulation (mostly just building dicts) is trivial and doesn't involve complex data processing.

While this may be much broader in some cases, it is also more standardized across platforms which I think makes it more valuable.

> it might be something we need to further tweak in the future.

+1 no objections to making improvements wherever needed